// @xene/core/dist/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const remove = (array, el) => {
    const index = array.indexOf(el);
    return index >= 0 ? array.splice(index, 1) : array;
};
class Channel {
    constructor() {
        this._managers = [];
        this._allManagers = [];
    }
    add(manager) {
        const canPrepare = manager.users.some(u => !this.hasFor(u));
        this._managers.push(manager);
        if (canPrepare)
            manager.prepare();
    }
    processMessage(message) {
        const manager = this._headFor(message.user);
        const canContinue = manager && manager.perform(message);
        if (canContinue)
            this._prepareNext(manager);
    }
    hasFor(user) {
        for (const { users } of this._managers)
            if (users.includes(user))
                return true;
        return false;
    }
    bind(manager) {
        this._allManagers.push(manager);
    }
    without(manager) {
        this._allManagers = this._allManagers.filter(m => m !== manager);
        this._managers = this._managers.filter(m => m !== manager);
    }
    abort(user) {
        const head = this._allManagers.find(m => m.users.includes(user));
        if (head)
            head.abort();
    }
    _prepareNext(manager) {
        const users = manager.users.reduce((acc, u) => this._headFor(u) === manager ? acc.concat(u) : acc, []);
        remove(this._managers, manager);
        users.forEach(u => this.hasFor(u) ? this._headFor(u).prepare() : null);
    }
    _headFor(user) {
        for (const manager of this._managers)
            if (manager.users.includes(user))
                return manager;
    }
}
exports.Channel = Channel;
