// @xene/core/dist/dialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const manager_1 = require("./manager");
const pause_1 = require("./actions/pause");
const parse_1 = require("./actions/parse");
const question_1 = require("./actions/question");
const isMessage = (arg) => arg != null && typeof arg !== 'function';
const errorHandler = (say, arg) => isMessage(arg) ? () => say(arg) : arg;
class Dialog {
    constructor(bot, channel, users) {
        this.isPaused = false;
        this.bot = bot, this.channel = channel, this.users = users;
        this._manager = new manager_1.Manager(bot, channel, users);
        this.parse = this.parse.bind(this);
        this.ask = this.ask.bind(this);
        this.say = this.say.bind(this);
    }
    get user() { return this.users[0]; }
    on(event, callback) {
        this._manager.on(event, callback);
    }
    end() {
        this._manager.emit('end');
        this._manager.unbind();
    }
    pause(message) {
        this.isPaused = true;
        this._manager.emit('pause');
        const handler = () => this.say(message, false);
        this._manager.add(new pause_1.Pause(handler));
    }
    say(message, unpause = true) {
        this._manager.emit('outgoingMessage', this.channel, message);
        if (this.isPaused && unpause) {
            this._manager.unpause();
            this._manager.emit('unpause');
            this.isPaused = false;
        }
        return this.bot.say(this.channel, message);
    }
    parse(parser, onError) {
        const parserObj = new parse_1.Parse(parser, errorHandler(this.say, onError));
        this._manager.add(parserObj);
        return parserObj.promise;
    }
    ask(message, parser, onError) {
        const sayMessage = () => this.say(message);
        const question = new question_1.Question(sayMessage, parser, errorHandler(this.say, onError));
        this._manager.add(question);
        return question.promise;
    }
}
exports.Dialog = Dialog;
