// @xene/core/dist/manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ee_1 = require("./ee");
const parse_1 = require("./actions/parse");
const pause_1 = require("./actions/pause");
const question_1 = require("./actions/question");
const isPause = (p) => p instanceof pause_1.Pause;
const isParse = (p) => p instanceof parse_1.Parse;
const isQuestion = (q) => q instanceof question_1.Question;
class Manager {
    constructor(_bot, _channelID, users) {
        this._bot = _bot;
        this._channelID = _channelID;
        this.users = users;
        /** @internal */
        this._ee = new ee_1.EventEmitter();
        this.emit = this._ee.emit;
        this.on = this._ee.on;
        this._queue = [];
        this._messages = [];
        this._channel.bind(this);
    }
    /**
     * For questions call to ask is required and so it
     * should be just added to the end of queue until
     * previous question/parsers finish.
     *
     * For Parser there no need to prepare it
     * and it's possible just to parse last message
     * since it may contain info parser tries to get
     * (unlike questions, which need to ask question
     * before parsing). So call to parse here will do
     * almost the same thing as in perform but without
     * triggering error hander of parser to not mess
     * up with queue. If it fails then same flow applied
     * as for question e.g. add to queue.
     */
    add(action) {
        if (isQuestion(action) || isPause(action)) {
            this._queue.push(action);
            return this._channel.add(this);
        }
        if (this._lastMessage && action.perform(this._lastMessage))
            return;
        else
            this._queue.push(action);
        if (!this._queue.some(isQuestion))
            this._channel.add(this);
    }
    prepare() {
        if (isQuestion(this._head))
            this._head.ask();
        else if (isPause(this._head))
            return;
        else
            this._head.failed(this._lastMessage);
    }
    perform(message) {
        this.emit('incomingMessage', message);
        this._messages.push(message);
        if (this._isEmpty)
            return true;
        if (!this._head.perform(message)) {
            this._head.failed(message);
            return false;
        }
        this._queue.shift();
        if (isQuestion(this._head))
            return true;
        else
            return this.perform(message);
    }
    unpause() {
        this._queue = this._queue.filter(p => !isPause(p));
    }
    abort() {
        this.emit('abort');
    }
    unbind() {
        this._channel.without(this);
    }
    get _lastMessage() { return this._messages[this._messages.length - 1]; }
    get _channel() { return this._bot._channelFor(this._channelID); }
    get _isEmpty() { return this._queue.length === 0; }
    get _head() { return this._queue[0]; }
}
exports.Manager = Manager;
