// @xene/slack/dist/slackbot.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const uuid = require("uuid");
const _ = require("lodash");
const core_1 = require("@xene/core");
const dispatcher_1 = require("./dispatcher");
const middleware_1 = require("./middleware");
const is_mentioned_1 = require("./helpers/is-mentioned");
const channel_type_1 = require("./helpers/channel-type");
const api_1 = require("./api");
class Slackbot extends core_1.Bot {
    constructor(options) {
        super();
        this.id = options.id || uuid.v4();
        this.chat = new api_1.Chat(options.botToken);
        this.rtm = new api_1.RTM(options.botToken);
        // Some of these API scopes' methods require additional
        // scopes which are defined only for apps and app tokens respectively
        this.auth = new api_1.Auth(options.appToken || options.botToken);
        this.users = new api_1.Users(options.appToken || options.botToken);
        this.groups = new api_1.Groups(options.appToken || options.botToken);
        this.channels = new api_1.Channels(options.appToken || options.botToken);
        this.files = new api_1.Files(options.appToken || options.botToken);
        if (options.dispatcher)
            options.dispatcher.add(this.id, this);
        else
            Slackbot.dispatcher.add(this.id, this);
    }
    say(channel, message) {
        return __awaiter(this, void 0, void 0, function* () {
            const init = { text: '', attachments: [] };
            message = _.isString(message) ? Object.assign({}, init, { text: message }) : Object.assign({}, init, message);
            message.attachments.forEach(a => a.callbackId = a.callbackId || this.id);
            return this.chat.postMessage(channel, message);
        });
    }
    listen() {
        this.rtm.on('message', this.onRtmMessage.bind(this));
        this.rtm.connect().then(i => this.bot = i.self);
        return this;
    }
    /**
     * Process incoming interactive messages
     * like button actions from slack
     * Called from Dispatcher
     */
    onInteractiveMessage(payload) {
        return __awaiter(this, void 0, void 0, function* () {
            const selected = payload.actions[0];
            const text = payload.originalMessage.text;
            let attachments = payload.originalMessage.attachments;
            attachments = attachments.map(this.markActionSelected.bind(this, selected));
            this.onMessage({
                channel: payload.channel.id,
                user: payload.user.id,
                text: selected.value,
                id: payload.ts
            });
            return { text, attachments };
        });
    }
    /**
     * Process new incoming RTM messages
     */
    onRtmMessage(payload) {
        const { user, ts, text, channel } = payload;
        if (this.bot.id === user)
            return;
        const isBotMentioned = is_mentioned_1.isMentioned(this.bot.id, text);
        const isPrivate = channel_type_1.isPrivateChannel(channel);
        if (!isPrivate && !isBotMentioned)
            return;
        this.onMessage({ id: ts, text, user, channel });
    }
    markActionSelected(action, attachment) {
        const selectedReplacer = ':white_check_mark: ' + action.text;
        if (_.find(attachment.actions, ['value', action.value])) {
            const title = attachment.title;
            delete attachment.actions;
            attachment.title = title ? (title + '\n' + selectedReplacer) : selectedReplacer;
        }
        return attachment;
    }
}
/**
 * Default dispatcher, used when user didn't provide
 * custom dispatcher. This is mostly used when user has
 * one type of bot, which is a common case
 */
Slackbot.dispatcher = new dispatcher_1.SlackbotDispatcher();
Slackbot.middleware = middleware_1.middleware;
Slackbot.oauthAccess = api_1.Auth.access;
exports.Slackbot = Slackbot;
