// @xene/slack/dist/middleware/middlewares.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const qs = require("querystring");
const rawBody = require("raw-body");
const request = require("request-promise-native");
const lodash_1 = require("lodash");
const case_1 = require("../helpers/case");
const format = require("../helpers/formatters/message");
const streamPayload = req => rawBody(req, { encoding: true }).then(qs.parse).then(i => JSON.parse(i.payload));
const existingPayload = payload => typeof payload === 'string' ? JSON.parse(payload) : payload;
const middlewareContext = (payload) => {
    const action = payload.actions[0];
    const { user, team, channel, token, callbackId, responseUrl } = payload;
    return {
        user, team, channel, token, callbackId, ephemeral: undefined, responseUrl,
        message: format.fromSlack(payload.originalMessage),
        action: {
            value: action.type === 'button' ? action.value : action.selectedOptions[0].value,
            type: action.type,
            id: action.name
        }
    };
};
const processRequestWithHandler = (handler, payload) => __awaiter(this, void 0, void 0, function* () {
    const context = middlewareContext(case_1.camel(payload));
    yield handler(context);
    let response;
    const { ephemeral, message } = context;
    const ephemeralAdded = ephemeral != null;
    const deleted = message == null;
    const modified = !deleted && !lodash_1.isEqual(message, format.fromSlack(payload.original_message));
    if (ephemeralAdded && modified)
        throw new Error("Can't show ephemeral message and update original message in the same time.");
    if (modified)
        response = format.toSlack(message);
    if (deleted)
        response = { delete_original: true };
    if (ephemeralAdded) {
        response = lodash_1.isString(ephemeral) ? { text: ephemeral } : format.toSlack(ephemeral);
        response.response_type = 'ephemeral';
        response.replace_original = false;
        response.delete_original = deleted;
    }
    if (!modified && !deleted && !ephemeralAdded)
        return;
    const body = response || payload.original_message;
    request.post({ uri: context.responseUrl, body, json: true });
});
exports.koa = (handler, ctx, next) => __awaiter(this, void 0, void 0, function* () {
    if (ctx.method.toLowerCase() !== 'post')
        return;
    // ctx.request.body hack is to get parsed body
    // if any body parsers middleware is already used in code
    // tslint:disable
    const payload = existingPayload(lodash_1.get(ctx, 'request.body.payload')) || (yield streamPayload(ctx.req));
    processRequestWithHandler(handler, payload);
    ctx.status = 200;
    ctx.body = '';
    return next();
});
exports.express = (handler, req, res, next) => __awaiter(this, void 0, void 0, function* () {
    if (req.method.toLowerCase() !== 'post')
        return;
    const payload = existingPayload(lodash_1.get(req, 'body.payload')) || (yield streamPayload(req));
    processRequestWithHandler(handler, payload);
    res.status(200).end();
    return next();
});
