// @xene/slack/dist/api/auth.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = require("./base");
const request_1 = require("./base/request");
const case_1 = require("../helpers/case");
const errors_1 = require("../errors");
class Auth extends base_1.APIModule {
    static access(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const uri = `https://slack.com/api/oauth.access`;
            const form = {
                client_id: options.id,
                client_secret: options.secret,
                code: options.code,
                redirect_uri: options.redirectUri
            };
            const response = yield request_1.request(uri, form);
            if (!response.ok)
                throw new errors_1.APIError(response.error);
            return case_1.camel(response);
        });
    }
}
exports.Auth = Auth;
