// @xene/slack/dist/api/channels.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fp_1 = require("lodash/fp");
const base_1 = require("./base");
const get_1 = require("../helpers/get");
class Channels extends base_1.APIModule {
    constructor() {
        super(...arguments);
        this.archive = (channel) => this.request('archive', { channel });
        this.create = (name, validate = false) => this.request('create', { name, validate }).then(get_1.get('channel'));
        this.history = (channel, options) => this.request('history', Object.assign({ channel }, options)).then(get_1.get('messages'));
        this.info = (channel) => this.request('info', { channel }).then(get_1.get('channel'));
        this.invite = (channel, user) => this.request('invite', { channel, user }).then(get_1.get('channel'));
        this.join = (channel) => this.request('join', { name: channel }).then(get_1.get('channel'));
        this.kick = (channel, user) => this.request('kick', { channel, user });
        this.leave = (channel) => this.request('leave', { channel });
        this.list = () => this.request('list').then(get_1.get('channels'));
        this.mark = (channel, ts) => this.request('mark', { channel, ts });
        this.rename = (channel, name, validate = true) => this.request('rename', { channel, name, validate }).then(get_1.get('channel'));
        this.replies = (channel, threadTs) => this.request('rename', { channel, threadTs }).then(get_1.get('messages'));
        this.setPurpose = (channel, purpose) => this.request('setPurpose', { channel, purpose });
        this.setTopic = (channel, topic) => this.request('setTopic', { channel, topic });
        this.unarchive = (channel) => this.request('unarchive', { channel });
        this.getMessage = (channel, ts) => this.history(channel, { inclusive: true, count: 1, latest: ts }).then(fp_1.head);
    }
}
exports.Channels = Channels;
