// @xene/slack/dist/api/rtm/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const WebSocket = require("ws");
const eventemitter3_1 = require("eventemitter3");
const base_1 = require("../base");
const logger_1 = require("../../logger");
const PING_INTERVAL = 5000;
const MAX_PONG_INTERVAL = 20000;
const boundPromise = () => {
    const result = {};
    result.promise = new Promise((resolve, reject) => {
        result.resolve = resolve;
        result.reject = reject;
    });
    return result;
};
class RTM extends base_1.APIModule {
    constructor(token) {
        super(token);
        this.inc = 1;
        this.ee = new eventemitter3_1.EventEmitter();
        this.lastPong = 0;
        this.connect = () => __awaiter(this, void 0, void 0, function* () {
            const promise = boundPromise();
            const response = yield this.request('connect');
            this.ws = new WebSocket(response.url);
            this.ws.on('message', this.emit.bind(this));
            this.ws.on('close', this.reconnect.bind(this));
            this.ws.on('open', () => promise.resolve(response));
            return promise.promise;
        });
        this.on = this.ee.addListener.bind(this.ee);
        this.off = this.ee.removeListener.bind(this.ee);
    }
    typing(channel) {
        this.wsSend({ type: 'typing', channel });
    }
    emit(msgString) {
        const msg = JSON.parse(msgString);
        if (msg.type === 'hello')
            this.handleHello();
        if (msg.type === 'pong')
            this.lastPong = Date.now();
        logger_1.logger.verbose('Incoming RTM message %s', msg.type);
        this.ee.emit(msg.subtype ? `${msg.type}.${msg.subtype}` : msg.type, msg);
    }
    handleHello() {
        this.lastPong = Date.now();
        if (this.pingTimer)
            clearInterval(this.pingTimer);
        this.pingTimer = setInterval(this.pingServer.bind(this), PING_INTERVAL);
    }
    reconnect() {
        logger_1.logger.verbose('Reconnecting to RTM API');
        this.disconnect();
        this.connect();
    }
    disconnect() {
        clearInterval(this.pingTimer);
        this.pingTimer = undefined;
        if (!this.ws)
            return;
        this.ws.removeAllListeners('close');
        this.ws.close();
    }
    pingServer() {
        const pongInterval = Math.abs(Date.now() - this.lastPong - PING_INTERVAL);
        if (pongInterval > MAX_PONG_INTERVAL)
            return this.reconnect();
        this.wsSend({ type: 'ping' });
    }
    wsSend(message) {
        this.ws.send(JSON.stringify(Object.assign({}, message, { id: this.inc })));
        this.inc += 1;
    }
}
exports.RTM = RTM;
